complex/ft_msg.c

```diff
@@ -290,6 +309,48 @@ int ft_recv_dgram(void)
 	return -FI_ETIMEDOUT;
 }
 
+int ft_recv_dgram_flood(void)
+{
+	struct timespec s, e;
+	int credits, ret;
+	int64_t poll_time = 0;
+	int expected = ft.xfer_iter;
+	int got	     = 0;
+	do {
+		if (ft_rx.credits > (ft_rx.max_credits >> 1)) {
+			ret = ft_post_recv_bufs();
+			if (ret)
+				return ret;
+		}
+
+		credits = ft_rx.credits;
+
+		ret = ft_comp_rx();
+		if (credits != ft_rx.credits) {
+			int num_creds = ft_rx.credits - credits;
+			poll_time=0;
+			got+=num_creds;
+		}
+		if (ret)
+			return ret;
+
+		if(got >= expected) {
+			return 0;
```

Comment:
I’ve seen a hang on ethernet network over the sockets provider when packets are lost, so I can confirm this is an issue.

I played around with sending a “done” special message, but the sends/receives are currently all from/to the same buffer (at least for the send variants, and get more complicated for the non-contigs).
This could lead to early termination or a hang if the messages are out of order.
There’s a couple ways to deal with this, but I’m not sure which is better:

1)  send immediate data.  This requires a code to use new completion formats, and new/modified tx_comp routines.
2)  use multiple receive buffers, so each buffer can be inspected individually for the “done” message.
3)  use another endpoint for out of band protocol messages

I suspect that we’ll need one or more of these mechanisms as we expand the tests.


> On Apr 8, 2015, at 2:54 PM, Sean Hefty <notifications@github.com> wrote:
> 
> In complex/ft_msg.c <https://github.com/ofiwg/fabtests/pull/221#discussion_r28017753>:
> 
> > +				return ret;
> > +		}
> > +
> > +		credits = ft_rx.credits;
> > +
> > +		ret = ft_comp_rx();
> > +		if (credits != ft_rx.credits) {
> > +			int num_creds = ft_rx.credits - credits;
> > +			poll_time=0;
> > +			got+=num_creds;
> > +		}
> > +		if (ret)
> > +			return ret;
> > +
> > +		if(got >= expected) {
> > +			return 0;
> The receive side does not exit until it receives the exact number of sends requested for the test. But the send side only sends that many. If any messages are lost, the result of the test is that the send side will send its last message repeatedly, on a timeout, until the receiver has received the correct number.
> 
> This will result in lower performance than is expected. Also, the sender only retries the last send 1000 times. Some of the tests will send millions of messages. If more than 1000 packets are dropped, the test will fail.
> 
> It would be better for the receiver to simple receive messages until it gets the 'done' message. The performance data would then be based on the number of packets actually received.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/ofiwg/fabtests/pull/221/files#r28017753>.
>